File: src/ados/services/ground_station/uplink/health.py
"""Cloud reachability probing for the uplink router.

A successful TCP connect to port 443 of the cloud relay is treated as a
strong proxy for full reachability of the Cloudflare-fronted Convex
endpoint. We do not run a full TLS handshake from stdlib so the
dependency footprint stays minimal.

The probe can be bound to a specific interface via SO_BINDTODEVICE so
that the test exercises the path the router currently selected, not
whatever the kernel default route happens to be at probe time.
"""

from __future__ import annotations

import asyncio
import socket
from typing import Optional

import structlog

__all__ = [
    "HEALTH_INTERVAL_SECONDS",
    "HEALTH_TIMEOUT_SECONDS",
    "HEALTH_HOST",
    "HEALTH_PORT",
    "HEALTH_PATH",
    "probe_host",
]

log = structlog.get_logger(__name__)

HEALTH_INTERVAL_SECONDS = 15.0
HEALTH_TIMEOUT_SECONDS = 5.0
HEALTH_HOST = "convex.altnautica.com"
HEALTH_PORT = 443
HEALTH_PATH = "/"


async def probe_host(iface: Optional[str]) -> bool:
    """TCP connect to the cloud relay, optionally bound to an iface.

    Returns True on a successful connect, False on DNS failure, timeout,
    or any socket error. SO_BINDTODEVICE requires CAP_NET_RAW. When the
    capability is missing we fall back to a plain connect on the current
    default route, which still validates reachability.
    """
    loop = asyncio.get_running_loop()
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.setblocking(False)
        if iface is not None:
            try:
                sock.setsockopt(
                    socket.SOL_SOCKET,
                    socket.SO_BINDTODEVICE,
                    iface.encode("ascii"),
                )
            except (PermissionError, OSError) as exc:
                log.debug(
                    "uplink.bind_iface_failed",
                    iface=iface,
                    error=str(exc),
                )
        try:
            addr_info = await loop.getaddrinfo(
                HEALTH_HOST, HEALTH_PORT, type=socket.SOCK_STREAM
            )
        except OSError as exc:
            log.debug("uplink.dns_failed", error=str(exc))
            sock.close()
            return False
        if not addr_info:
            sock.close()
            return False
        family, socktype, proto, _, sockaddr = addr_info[0]
        if family != socket.AF_INET:
            sock.close()
            sock = socket.socket(family, socktype, proto)
            sock.setblocking(False)
        try:
            await asyncio.wait_for(
                loop.sock_connect(sock, sockaddr),
                timeout=HEALTH_TIMEOUT_SECONDS,
            )
            return True
        except (asyncio.TimeoutError, OSError) as exc:
            log.debug("uplink.connect_failed", iface=iface, error=str(exc))
            return False
        finally:
            try:
                sock.close()
            except OSError:
                pass
    except Exception as exc:
        log.debug("uplink.probe_exc", error=str(exc))
        return False

File: crates/ados-net/src/router/health.rs
//! Cloud-reachability probing for the uplink router.
//!
//! A successful TCP connect to port 443 of the cloud relay is treated as a
//! strong proxy for reachability of the Cloudflare-fronted endpoint. We do not
//! run a TLS handshake, keeping the dependency footprint minimal (no crypto).
//!
//! The probe can be bound to a specific interface via `SO_BINDTODEVICE` so the
//! test exercises the path the router currently selected, not whatever the
//! kernel default route happens to be at probe time. Binding needs
//! `CAP_NET_RAW`; on `EPERM` (or any non-Linux host where the option does not
//! exist) we fall back to a plain connect on the current default route, which
//! still validates reachability. Ports `uplink/health.py`.

use std::net::ToSocketAddrs;
use std::time::Duration;

use socket2::{Domain, Protocol, SockAddr, Socket, Type};
use tracing::debug;

/// Health-check cadence (the router's tick interval).
pub const HEALTH_INTERVAL: Duration = Duration::from_secs(15);
/// Per-connect timeout.
pub const HEALTH_TIMEOUT: Duration = Duration::from_secs(5);
/// Cloud relay host.
pub const HEALTH_HOST: &str = "convex.altnautica.com";
/// Cloud relay port (HTTPS).
pub const HEALTH_PORT: u16 = 443;
/// Probe path (informational; no HTTP request is sent, connect only).
pub const HEALTH_PATH: &str = "/";

/// TCP-connect to the cloud relay, optionally bound to `iface`. Returns `true`
/// on a successful connect, `false` on DNS failure, timeout, or any socket
/// error. Runs the blocking resolve + connect on a tokio blocking thread.
pub async fn probe_host(iface: Option<&str>) -> bool {
    let iface = iface.map(|s| s.to_string());
    match tokio::task::spawn_blocking(move || probe_blocking(iface.as_deref())).await {
        Ok(ok) => ok,
        Err(exc) => {
            debug!(error = %exc, "uplink.probe_exc");
            false
        }
    }
}

/// Blocking probe body. Resolves the host, creates a socket, optionally binds
/// it to the interface, and connects with [`HEALTH_TIMEOUT`].
fn probe_blocking(iface: Option<&str>) -> bool {
    // getaddrinfo equivalent; first A/AAAA record wins, matching the Python
    // `addr_info[0]`. DNS failure → unreachable.
    let target = format!("{HEALTH_HOST}:{HEALTH_PORT}");
    let sockaddr = match target.to_socket_addrs() {
        Ok(mut it) => match it.next() {
            Some(addr) => addr,
            None => {
                debug!("uplink.dns_failed");
                return false;
            }
        },
        Err(exc) => {
            debug!(error = %exc, "uplink.dns_failed");
            return false;
        }
    };

    let domain = Domain::for_address(sockaddr);
    let sock = match Socket::new(domain, Type::STREAM, Some(Protocol::TCP)) {
        Ok(s) => s,
        Err(exc) => {
            debug!(error = %exc, "uplink.probe_exc");
            return false;
        }
    };

    // SO_BINDTODEVICE: Linux-only and capability-gated. EPERM (or any error)
    // logs at debug and proceeds with a plain connect on the default route.
    if let Some(name) = iface {
        bind_device_best_effort(&sock, name);
    }

    match sock.connect_timeout(&SockAddr::from(sockaddr), HEALTH_TIMEOUT) {
        Ok(()) => true,
        Err(exc) => {
            debug!(iface = iface, error = %exc, "uplink.connect_failed");
            false
        }
    }
}

#[cfg(any(target_os = "android", target_os = "fuchsia", target_os = "linux"))]
fn bind_device_best_effort(sock: &Socket, iface: &str) {
    if let Err(exc) = sock.bind_device(Some(iface.as_bytes())) {
        debug!(iface = iface, error = %exc, "uplink.bind_iface_failed");
    }
}

#[cfg(not(any(target_os = "android", target_os = "fuchsia", target_os = "linux")))]
fn bind_device_best_effort(_sock: &Socket, iface: &str) {
    // SO_BINDTODEVICE does not exist off Linux; the Python path also degrades
    // to a plain connect here.
    debug!(iface = iface, "uplink.bind_iface_unsupported");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn constants_match_python() {
        assert_eq!(HEALTH_INTERVAL, Duration::from_secs(15));
        assert_eq!(HEALTH_TIMEOUT, Duration::from_secs(5));
        assert_eq!(HEALTH_HOST, "convex.altnautica.com");
        assert_eq!(HEALTH_PORT, 443);
        assert_eq!(HEALTH_PATH, "/");
    }

    #[tokio::test]
    async fn probe_against_a_dead_port_is_false() {
        // 127.0.0.1 with a closed port refuses fast → false, exercises the
        // connect-error path without network access. We can't point probe_host
        // at it (host is fixed), so drive probe_blocking via a local override
        // is not possible; instead assert a probe to a bound-but-unroutable
        // iface name fails gracefully (bind error → plain connect → may try
        // the real host, which in CI has no network → false). Kept hermetic by
        // only asserting it returns a bool without panicking.
        let _ = probe_host(Some("ados-nonexistent-iface0")).await;
    }
}
